### DISCUSSION TITLE
How to get more direct turn by turn routes?

### DISCUSSION THREAD

**Discussion by localjo** [osmandapp/OsmAnd#11044] [Q&A] [answered]
Hello, I'm new to OsmAnd. Thank you so much to the maintainers and contributors for this project. I'd love to contribute someday myself. :) 

I've just started using OsmAnd, and I'm in the process of tweaking all of the settings to match my preferences. I'm having trouble figuring out how to tweak settings related to routing. I'm finding that, by default, OsmAnd is routing me through a lot of back roads with far more turns than necessary. Often there is a very clear main road that I can take to my destination and OsmAnd routes me through side streets in what seems to me to be a very inefficient route.

I'm curious if there are any settings I can tweak to get better routes? I saw a setting that says "Use 2-phase A* routing algorithm", and tried toggling it on and off but didn't see a difference in the resulting routes.

Thanks for your help!

**Comment by vshcherb** [comment]
Probably it's not possible via UI and you need to get a finer control over the app. So you can enable your routing profile, please check https://github.com/osmandapp/OsmAnd-resources/blob/master/routing/routing.xml (it consists of section for different routing profiles), you can tweak some parameters and import later to OsmAnd

**Comment by barteis** [comment]
Just curious: when you press the navigation button, you see a set of icons indicating different navigation types.  Which one are you using?

**Reply by localjo** [reply]
This is in driving mode.

**Reply by barteis** [reply]
Under options (the gear) Fuel-efficient way disabled? Avoid roads options all disabled?  No Avoid roads markers inadvertently set? I know these are probably obvious, but i’ve not seen this behavior in driving mode.

**Reply by localjo** [reply]
Yes, all of those are disabled.

**Comment by jfburke3** [comment]
My first question relates to the various attributes on those roads in OSM itself.  Are speed limits set correctly?  Are there stop signs or traffic signals set on that main road but not on the secondary roads?  There are many items in OSM that will affect how OsmAnd chooses a route.

**Reply by localjo** [reply]
I'm not sure how to check that. Here's an example of a route that should be just two turns (red line) but is instead ~14 turns. In Budapest from Ferenciek Tere to Hosok Tere. 

![photo_2021-03-04_00-17-57](https://user-images.githubusercontent.com/516388/109886049-cd2e8180-7c7f-11eb-8aa1-876372080de6.jpg)

How can I check the underlying OSM data? And if there is a problem with the data, is there a setting I can tweak to get better routes with existing data? Or is the only solution to fix the maps of all of the side streets in Budapest?

**Reply by barteis** [reply]
I downloaded the map, tried your route, and got a somewhat similar result, but with 8 turns rather than 14. When I added an intermediate destination to force a left turn on Karoly Korbut, as I think you have it, OsmAnd routed right, and then a u-turn.  Indicating perhaps that the left isn't possible.  Or at least the OSM data doesn't allow it.  

 You can check the underlying OSM data on the OpenStreetMap website, and if you register, can correct mapping errors.  Tweaking OsmAnd routing is beyond my abilities, but see vhscherb's note.

**Comment by tommycrock** [comment]
I got a more straightforward route from graphHopper and OSRM
[image: image.png]
https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=47.4920%2C19.0526%3B47.5142%2C19.0768#map=15/47.5028/19.0619

There is no left turn (at least in OSM) where you wanted to take one:
[image: image.png]

But I don't know why the turns are so confused in yours around Karoly
Korut. I'd consider filing a routing report on github
https://github.com/osmandapp/OsmAnd/issues/new?assignees=&labels=&template=3-routing-report.md

**Reply by localjo** [reply]
Yes, the route you got from graphHopper and OSRM is much better. (The left turn in question is because you put the starting point in a slightly different spot, but no worries).

Anyway, isn't there some way to get these OSRM directions in OsmAnd? This isn't just happening with one route, but almost any time I try to navigate somewhere. Is it possible to configure which routing service OsmAnd uses?

**Reply by localjo** [reply]
Ah! I found the setting to change the routing engine. It was a bit hidden, but I'm able to get the routes I need now.

**Comment by localjo** [comment, marked as answer]
I found the answer to my question. It is possible to add routing engines to OsmAnd and change the routing engine that OsmAnd uses for directions, for each profile in the app. Here's how:

- Go to the settings for the profile you want to change the routing engine on. (Settings > App Profiles > [Profile Name])
- Open the navigation type settings (Navigation settings > Navigation type)
- Scroll to the bottom and tap "Add online routing engine"
- Choose the engine you want to add and tap "Save"
- Tap on "Navigation type " again, and select the routing engine you want to use from the list

It's probably related to the OSM map quality in my area, but I had better quality routes with the Grasshopper and OSRM routing engines.

**Reply by daslondon** [reply]
You have raised a very important point.  So far I have used OsmAnd in southern England and in London.
If I were to come to Budapest and wish to navigate, say, between the two points you mention, how would I know that the directions given are not optimal?
Do I use another navigation app to compare?  Waste of time, I think.

I have to say that on my Windows Phone I was very happy with HERE Maps (which later evolved into the Microsoft Maps fork on the phone).  When I abandoned that OS at the end of 2019 and switched to Android, I discovered OsmAnd, which has an attractive UI.
No surprise I installed the HERE WeGo and Google Maps apps as well.
The special beauty of HERE WeGo is free unlimited maps for offline navigation...

**Comment by navid-zamani** [comment]
This is still a problem as of today.
The whole point of using OSMand is that it’s offline. I ride 8 hour tours. Even if I had reception everywhere in the boondocks, the phone simply doesn’t last that long with full GPS and wifi and screen on in the hot sun. ^^

OSMand literally goes straight along the obvious street… then for no reason at all, changes to a parallel street, and later back.
I know it is for no reason, because I ride the street multiple times a week.
And no matter what I do, it *will not* use a 10m train track crossing or a literal 2m from a cul de sac to a street that’s just separated by some sidewalk.

It would be half as bad, if I could at least edit the track to tell it “go through here, no matter what”. But instead, it navigates the wrong path, turns around, into the path I told it, turns around again, and *then* continues the route.
I even tried exporting it as a GPX track, in the hope that I could just edit the vertices. But no, somehow it only lets me edit the start and end point, and between that there are apparently no points, and all it offers me is to recalculate or choose a different routing profile. No “manual” editing… as in, the whole point of having a GPX track. \*bangs head on table\*
Sure, I can add points to the *end*. But “there is no spoon… err, … in-between” .

Together with the extremely annoying zoom, that literally always zooms to the wrong place, so zooming in and out a few times and suddenly you’re in Scandinavia! … I really went from an OSMand lover to hating it with a passion. (The funniest part is that you *really* try to force it, it just crashes on you. Multiple times. XD)
I don’t want to hate OSMand. Going back to GMaps after coming from it feels like “going back to Playboy”. (Our city officially uses OpenStreetMap and puts pretty much literally every trash can and tree into it, so I may be spoiled.)  XD

**I just want to be in control of my own damn things! Edit the track however I like.**
That’s all I ask!
Not some condescending Google-Apple-style uber-bro that always “knows best” and doesn’t let you have your own breath by yourself if wasn’t the one commanding you to do it.

**So:** Can you *please* just make the automatic route finder navigation thingie a tool that merely creates GPX tracks, with every turn (vertex) being a new point, and the ability to insert points easily, like you do in the navigation thingie. Instead of it being somehow two separate rabbit holes with 90% overlapping (duplicated) functionality. I mean saving a route as a track is already possible, even if well-hidden. :/ You’d save your self so much work just from losing the duplications!

**Reply by 0bWAN** [reply]
OMG, is your brain somehow connected to mine? As I read this post, I wondered if I had written it and forgotten. Well said! Well written. Obviously you are a highly intelligent individual! ;)
